content/blog/package-react-components/index.md

```diff
@@ -0,0 +1,776 @@
+
+### Active Compilation
+
+#### As Component Authors
+
+You actually need to write a full web pack configuration
+to generate the bundle, but it is really not that bad (more on this later).
+
+Alternatively, if you are OK with Not-Inlining assets then
+you can use already made solutions such as [create-react-library][105] et al.
+
+You will need to make decisions about how to package and distribute your
+code which _is_ exactly the main point of this post.
+
+#### As Component Consumers
+
+You don't need to customize your Webpack config to use the library,
+this results in simpler configuration. You can also use standard
+pre-made solutions such as [create-react-app][16]
+
+**Theming** will depend exclusively on what the _Component Author_
+sets up to be the Theming API. If a Component Author use
+something like **CSS-in-JS** then the experience will be pretty good.
+[Material-UI React][7] uses this API.
+
+If the Component Author provides a build to compile custom Themes
+out of band (outside your Apps build system) then it will be less
+optimal in my opinion.
+[Ant Design][8] and [React Bootstrap][9] use this API.
+
+## Theming
+
+For Component Libraries or UI Frameworks **Theming** is a _must_. They need to provide a way
+for Component Consumers to change the default look of the components to fit
+their brand or their business.
+
+Lets inspect different strategies of providing Theming and Customization hooks.
+
+### Regular CSS Class Names
+
+This approach works great for single or few component libraries but falls
+very short for larger Component Libraries or UI Frameworks.
+
+The basic premise is that Consumers use well defined CSS class names
+to alter the Component looks.
+
+This approach works only at **Run-Time**.
+
+#### As Component Author
+
+- You write regular CSS Classes (without CSS modules) and you relate them to your components via
+  `className`.
+- You can use LESS or SASS or any other CSS preprocessor.
+- You should use Class Names naming conventions such as [BEM][111] or [SMACSS][112] among others.
+- You **cannot** use CSS Modules since they generate pseudo random Class Names that cannot be used consistently by Consumers to customize or theme your components.
+- You structure your classes carefully to allow Component Consumers to Theme or Customize your components styles by overriding some of those classes.
+
+#### As Component Consumer
+
+You use the Component Author provided class names to alter the component looks.
+
+### CSS Theme variables
+
+This approach is much more robust. The main premise is that Component Authors
+use a set of well defined CSS, LESS or SASS variables to derive the rest of the styles.
+These variables conform the Theme. Component Consumers can alter these variables
+to generate different looking themes.
+
+Examples of these are
+
+- [AntDesing][8]
+- [Bootstrap][9]
+- [Semantic-UI][14]
+
+This approach works only at **Build-Time** and **may require a build system**.
```

Comment:
The "May require" comes from the fact that CSS variables do not require a build system, whereas LESS and SASS variables do require some build system to work?

Reply:
yush

Reply:
Added clarification and links